### PR TITLE
Fix some issues

### PR DESCRIPTION
**Pull request by takeharukato**
Hi, tyfkda,

As I posted on your issue tracker, I found three fixes were needed.
I tried to fix them.
Would you please check them and merge these fixes into you repository if you would like to do it.

Thanks, 
Regards,

### CODE REVIEW

**Comment by tyfkda** [comment]
I've merged c775a5df5090f03a47ce096b07beea73dbe6a360 .

I will look into other commits.

File: kernel/fs.c [hunk 1]
```diff
@@ -402,41 +402,122 @@ bmap(struct inode *ip, uint bn)
...
+
+  if ( soff > 0 ) {
+
+    bp = bread(ip->dev, bmap(ip, sblk));
+    memset(bp->data + soff, 0, BSIZE-soff);
+    log_write(bp);
+    brelse(bp);
+    ++sblk;
+  }
+  for( cur = sblk; eblk >= cur; ++cur) {
```
Comment (tyfkda) [inline kernel/fs.c:478]: Does this loop work as intended?

This function is called only when `new <= ip->size`, so `sblk >= eblk`.
So the loop never happens and no block released.

File: kernel/fs.c [hunk 2]
```diff
@@ -402,41 +402,122 @@ bmap(struct inode *ip, uint bn)
...
 
-  if(ip->addrs[NDIRECT]){
-    bp = bread(ip->dev, ip->addrs[NDIRECT]);
-    a = (uint*)bp->data;
-    for(j = 0; j < NINDIRECT; j++){
-      if(a[j])
-        bfree(ip->dev, a[j]);
-    }
-    brelse(bp);
+  if ( eblk < NDIRECT ) {
```
Comment (tyfkda) [inline kernel/fs.c]: It should be checked whether indirect address is used
to prevent freeing 0 block.

File: kernel/fs.c (hunk 1 above)
Comment (takeharukato) [reply kernel/fs.c:478]: this loop is needed if new < ip->size and new indicate indirect blocks, isn't it?

Comment (tyfkda) [reply kernel/fs.c:478]: I mean it is needed to swap start and end values each other:

```c
  for( cur = eblk; sblk >= cur; ++cur) {
```

**Comment by takeharukato** [comment]
> I mean it is needed to swap start and end values each other:

Right.
But sblk means the start block and eblk means the end of blocks, so I think that it is more preferable to modify assignments of each variable.

Thanks,
Regards,

**Comment by takeharukato** [comment]
Hi,

Thank you for your careful review.
Sorry to my mistake.

Anyway,  I've  pushed a fix a bug which is mentioned above.

Thanks,
Regards,

File: kernel/fs.c [hunk 3]
```diff
@@ -459,10 +459,11 @@ free_space_inode(struct inode *ip, uint new){
...
-  sblk = ip->size / BSIZE; 
-  soff = ip->size % BSIZE;
-  eblk = new / BSIZE; 
-  eoff = new % BSIZE;
+  sblk = new / BSIZE; 
+  soff = new % BSIZE;
+  eblk = ip->size / BSIZE; 
+  eoff = ip->size % BSIZE;
+
   if ( eoff > 0 )
```
Comment (tyfkda) [inline kernel/fs.c]: You have to replace all variable usage below.

Comment (tyfkda) [reply kernel/fs.c]: I misunderstood the usage.
Forget about it.

But I think [line 501](https://github.com/tyfkda/xv6/pull/21/commits/4d2ba765686c375e6e8f83478cf259fb2b896e51#diff-c7124660ca155a8c4b0b37f9a9d06167L500) should be modified:
```diff
-  if ( eblk < NDIRECT ) {
+  if ( sblk <= NDIRECT && ip->addrs[NDIRECT]) {  // Edge case: when soff > 0 then ++sblk, so sblk == NDIRECT means indirect address is not used.
```

**Comment by takeharukato** [comment]
Hi,

Oh, I could not notice that, Great!

```
-  if ( eblk < NDIRECT ) {
+  if ( sblk <= NDIRECT && ip->addrs[NDIRECT]) {  // Edge case: when soff > 0 then ++sblk, s
```

Right. 
I wrote the fix for this and push it.
Would you please confirm the fix.

Thanks,
Regards,

File: kernel/fs.c [hunk 4]
```diff
@@ -402,41 +402,128 @@ bmap(struct inode *ip, uint bn)
...
+        ip->addrs[cur] = 0;
+      }
+    }  else {
+    
+      if(ip->addrs[NDIRECT] != 0) {
+        
+        bp = bread(ip->dev, ip->addrs[NDIRECT]);
+        a = (uint*)bp->data;
+        
+        curidx= (cur - NDIRECT) / BSIZE;
```
Comment (tyfkda) [inline kernel/fs.c]: Dividing by `BSIZE` collapses index.

File: kernel/fs.c [hunk 5]
```diff
@@ -498,7 +498,12 @@ free_space_inode(struct inode *ip, uint new){
     }
   }
 
-  if ( eblk < NDIRECT ) {
+  /* Release an indirect index block.
+   * Edge case: Eblk == NDIRECT always means an indirect index block is no longer used
+   *            because we've already increase sblk above if eoff > 0.
+   */
+  if ( ( eblk <= NDIRECT ) && ( ip->addrs[NDIRECT] != 0 ) ) {
```
Comment (tyfkda) [inline kernel/fs.c]: I think this code doesn't work as expected.

Would you please confirm the behavior?

File: kernel/fs.c [hunk 6]
```diff
@@ -402,41 +402,128 @@ bmap(struct inode *ip, uint bn)
...
+    }  else {
+    
+      if(ip->addrs[NDIRECT] != 0) {
+        
+        bp = bread(ip->dev, ip->addrs[NDIRECT]);
+        a = (uint*)bp->data;
+        
+        curidx= (cur - NDIRECT) / BSIZE;
+        if(a[curidx] != 0)
+          bfree(ip->dev, a[curidx]);
```
Comment (tyfkda) [inline kernel/fs.c]: Also, you need to clear`a[curidx]` and write back the buffer.

**Comment by takeharukato** [comment]
> I think this code doesn't work as expected.

I was a little confuse when I wrote previous patch.
In this case, we should have checked sblk because sblk indicates the first remaining block as you said.

> Also, you need to cleara[curidx] and write back the buffer.

On second thought, I notice that I need to clear an entry of array a as you said.

> Dividing by BSIZE collapses index.

Sure. You are right, I'll fix it.

All fixes I mentioned above have been pushed.
Would you please check them?

Thanks,
Regards,

**Comment by tyfkda** [comment]
@takeharukato , thank you for the changes.

I've merged your commites about `itrunc`
(with cherry-pick and squash).
They solved #16 .

Thank you for your contribution!

**Comment by takeharukato** [comment]
You are welcome.
Thank you for your thorough review.

I should have review it by myself carefully.

**Pull request closed without merging**